sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala

```diff
@@ -237,21 +237,26 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
           new Path(metastoreRelation.catalogTable.storage.locationUri.get),
           partitionSpec)
 
-        val inferredSchema = if (fileType.equals("parquet")) {
-          val inferredSchema =
-            defaultSource.inferSchema(sparkSession, options, fileCatalog.allFiles())
-          inferredSchema.map { inferred =>
-            ParquetFileFormat.mergeMetastoreParquetSchema(metastoreSchema, inferred)
-          }.getOrElse(metastoreSchema)
-        } else {
-          defaultSource.inferSchema(sparkSession, options, fileCatalog.allFiles()).get
+        val inferredSchema =
+          defaultSource.inferSchema(sparkSession, options, fileCatalog.allFiles())
+        val schema = fileType match {
+          case "parquet" =>
+            // For Parquet, get correct schema by merging Metastore schema data types
+            // and Parquet schema field names.
+            inferredSchema.map { schema =>
+              ParquetFileFormat.mergeMetastoreParquetSchema(metastoreSchema, schema)
+            }.getOrElse(metastoreSchema)
+          case "orc" =>
+            inferredSchema.getOrElse(metastoreSchema)
+          case _ =>
+            inferredSchema.get
```

Comment:
@gatorsmile has asserted that we shouldn't get here. So shouldn't we throw some kind of exception here?